File: android/lca/src/main/java/com/dreampany/lca/vm/NotifyViewModel.java
package com.dreampany.lca.vm;

import android.app.Application;

import com.dreampany.framework.api.notify.NotifyManager;
import com.dreampany.framework.app.BaseApp;
import com.dreampany.framework.misc.AppExecutors;
import com.dreampany.framework.misc.ResponseMapper;
import com.dreampany.framework.misc.RxMapper;
import com.dreampany.framework.misc.exceptions.EmptyException;
import com.dreampany.framework.util.DataUtil;
import com.dreampany.framework.util.TextUtil;
import com.dreampany.framework.util.TimeUtil;
import com.dreampany.lca.R;
import com.dreampany.lca.app.App;
import com.dreampany.lca.data.enums.CoinSource;
import com.dreampany.lca.data.enums.Currency;
import com.dreampany.lca.data.model.Coin;
import com.dreampany.lca.data.model.CoinAlert;
import com.dreampany.lca.data.model.News;
import com.dreampany.lca.data.model.Price;
import com.dreampany.lca.data.model.Quote;
import com.dreampany.lca.data.source.pref.Pref;
import com.dreampany.lca.data.source.repository.ApiRepository;
import com.dreampany.lca.data.source.repository.CoinAlertRepository;
import com.dreampany.lca.data.source.repository.CoinRepository;
import com.dreampany.lca.data.source.repository.NewsRepository;
import com.dreampany.lca.data.source.repository.PriceRepository;
import com.dreampany.lca.misc.Constants;
import com.dreampany.lca.misc.CurrencyFormatter;
import com.dreampany.lca.ui.activity.HomeActivity;
import com.dreampany.lca.ui.enums.UiSubtype;
import com.dreampany.lca.ui.enums.UiType;
import com.dreampany.lca.ui.model.CoinAlertItem;
import com.dreampany.lca.ui.model.CoinItem;
import com.dreampany.lca.ui.model.NewsItem;
import com.dreampany.lca.ui.model.UiTask;
import com.dreampany.network.manager.NetworkManager;
import com.google.common.collect.Maps;

import java.util.List;
import java.util.Map;

import javax.inject.Inject;
import javax.inject.Singleton;
import io.reactivex.Flowable;
import io.reactivex.Maybe;
import io.reactivex.MaybeSource;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Function;
import timber.log.Timber;

/**
 * Created by Hawladar Roman on 7/22/2018.
 * Dreampany Ltd
 * dreampanymail@gmail.com
 */
@Singleton
public class NotifyViewModel {

    private final Application application;
    private final RxMapper rx;
    private final Pref pref;
    private final ApiRepository repo;
    private final CoinRepository coinRepo;
    private final PriceRepository priceRepo;
    private final CoinAlertRepository alertRepo;
    private final NewsRepository newsRepo;
    private final CurrencyFormatter formatter;
    private final NotifyManager notify;

    private final Map<Coin, Price> prices;

    @Inject
    NotifyViewModel(Application application,
                    RxMapper rx,
                    AppExecutors ex,
                    ResponseMapper rm,
                    NetworkManager network,
                    Pref pref,
                    ApiRepository repo,
                    CoinRepository coinRepo,
                    PriceRepository priceRepo,
                    CoinAlertRepository alertRepo,
                    NewsRepository newsRepo,
                    CurrencyFormatter formatter) {
        this.application = application;
        this.rx = rx;
        this.pref = pref;
        this.repo = repo;
        this.coinRepo = coinRepo;
        this.priceRepo = priceRepo;
        this.alertRepo = alertRepo;
        this.newsRepo = newsRepo;
        this.formatter = formatter;
        this.notify = new NotifyManager(application);
        prices = Maps.newConcurrentMap();
    }

    public BaseApp getApp() {
        return (BaseApp) application;
    }

    public void clear() {
        //disposables.dispose();
    }

    public void notifyIf() {
        Timber.v("Job Processing");
        notifyProfitableCoin();
        notifyCoin();
        notifyNews();
    }

    private void notifyProfitableCoin() {
        if (!pref.hasNotifyCoin()) {
            return;
        }
        if (!TimeUtil.isExpired(pref.getAlertProfitableCoin(), Constants.Delay.INSTANCE.getAlertProfitableCoinMS())) {
            return;
        }
        pref.commitAlertProfitableCoin();
        Timber.e("Fire Profitable Notification");
        Currency currency = pref.getCurrency(Currency.USD);
        Maybe<List<CoinItem>> profitMaybe = getProfitableItemsRx(currency);
        if (profitMaybe != null) {
            Disposable disposable = rx
                    .backToMain(profitMaybe)
                    .subscribe(result -> postProfitableCoins(currency, result), this::postFailed);
        } else {
            Timber.e("getProfitableItemsRx is Null");
        }
    }

    private void notifyCoin() {
        if (!pref.hasNotifyCoin()) {
            return;
        }
        if (!TimeUtil.isExpired(pref.getAlertCoin(), Constants.Delay.INSTANCE.getAlertCoinMS())) {
            return;
        }
        pref.commitAlertCoin();
        Timber.e("Fire Alert Notification");
        Currency currency = pref.getCurrency(Currency.USD);
        Maybe<List<CoinAlertItem>> alertMaybe = getAlertItemsRx(currency);
        if (alertMaybe != null) {
            Disposable disposable = rx
                    .backToMain(alertMaybe)
                    .subscribe(this::postResultAlerts, this::postFailed);
        } else {
            Timber.e("getAlertItemsRx is Null");
        }
    }

    private void notifyNews() {
        if (!pref.hasNotifyNews()) {
            return;
        }
        if (!TimeUtil.isExpired(pref.getAlertNews(), Constants.Delay.INSTANCE.getAlertNewsMS())) {
            return;
        }
        pref.commitAlertNews();
        Timber.e("Fire Notification");
        Maybe<List<NewsItem>> newsMaybe = getNewsItemsRx();
        if (newsMaybe != null) {
            Disposable disposable = rx
                    .backToMain(newsMaybe)
                    .subscribe(this::postResultNews, this::postFailed);
        } else {
            Timber.e("getNewsItemsRx is Null");
        }
    }

    private Maybe<List<CoinItem>> getProfitableItemsRx(Currency currency) {
        return Maybe.create(emitter -> {
            List<CoinItem> result =
                    coinRepo.getRandomItemsRx(CoinSource.CMC, currency, Constants.Limit.COIN_RANDOM)
                            .flatMap((Function<List<Coin>, MaybeSource<List<CoinItem>>>) coins -> getProfitableItemsRx(currency, coins))
                            .blockingGet();
/*            int resultMax = coinCount > Constants.Limit.COIN_PAGE ? coinCount : Constants.Limit.COIN_PAGE;

            int listStart = (resultMax == Constants.Limit.COIN_PAGE) ? 0 : NumberUtil.nextRand((resultMax - Constants.Limit.COIN_PAGE) + 1);
            int listLimit = Constants.Limit.COIN_PAGE;
            List<CoinItem> result = repo
                    .getItemsIfRx(CoinSource.CMC, currency, listStart, listLimit)
                    .flatMap((Function<List<Coin>, MaybeSource<List<CoinItem>>>) coins -> getProfitableItemsRx(currency, coins))
                    .blockingGet();*/

            if (emitter.isDisposed()) {
                return;
            }
            if (DataUtil.isEmpty(result)) {
                emitter.onError(new EmptyException());
            } else {
                emitter.onSuccess(result);
            }
        });
    }

    private Maybe<List<CoinAlertItem>> getAlertItemsRx(Currency currency) {
        return alertRepo
                .getItemsRx()
                .flatMap((Function<List<CoinAlert>, MaybeSource<List<CoinAlertItem>>>) alerts -> getAlertItemsRx(currency, alerts));
    }

    private Maybe<List<NewsItem>> getNewsItemsRx() {
        return newsRepo
                .getItemsRx(1)
                .flatMap((Function<List<News>, MaybeSource<List<NewsItem>>>) this::getNewsItemsRx);
    }

    private Maybe<List<CoinItem>> getProfitableItemsRx(Currency currency, List<Coin> result) {
        return Flowable.fromIterable(result)
                .filter(this::isProfitable)
                .map(coin -> {
                    CoinItem item = CoinItem.Companion.getSimpleItem(coin, currency, formatter);
                    return item;
                })
                .toList()
                .toMaybe();
    }

    private Maybe<List<CoinAlertItem>> getAlertItemsRx(Currency currency, List<CoinAlert> result) {
        return Flowable.fromIterable(result)
                .filter(alert -> isAlertable(currency, alert))
                .map(alert -> {
                    Coin coin = repo.getItemIf(CoinSource.CMC, currency, alert.getId());
                    return CoinAlertItem.getItem(coin, alert);
                }).toList()
                .toMaybe();
    }

    private Maybe<List<NewsItem>> getNewsItemsRx(List<News> result) {
        return Flowable.fromIterable(result)
                .map(NewsItem.Companion::getItem)
                .toList()
                .toMaybe();
    }

    private void postProfitableCoins(Currency currency, List<CoinItem> items) {
        App app = (App) application;
        if (app.isVisible()) {
            //return;
        }

        String title = TextUtil.getString(app, R.string.notify_title_profit);
        String message;
        Class<?> targetClass = null;
        UiTask<Coin> task = null;

        if (!DataUtil.isEmpty(items)) {
            CoinItem profitable = items.get(0);
            for (int index = 1; index < items.size(); index++) {
                Quote profitableQuote = profitable.getItem().getQuote(currency);
                Quote nextQuote = items.get(index).getItem().getQuote(currency);
                if (nextQuote.getDayChange() > profitableQuote.getDayChange()) {
                    profitable = items.get(index);
                }
            }
            Coin coin = profitable.getItem();
            double price = coin.getQuote(currency).getPrice();
            double dayChange = coin.getQuote(currency).getDayChange();
            message = formatter.formatPrice(coin.getSymbol(), coin.getName(), price, dayChange, currency);

            targetClass = HomeActivity.class;
            task = new UiTask<>(false,UiType.COIN, UiSubtype.VIEW, coin, null);
        } else {
            message = TextUtil.getString(app, R.string.profitable_coins_motto);
        }
        notify.showNotification(title, message, R.drawable.ic_notification, targetClass, task);
        getApp().throwAnalytics(Constants.Event.NOTIFICATION, Constants.Companion.notifyProfitableCoin(application));
    }

    private void postResultAlerts(List<CoinAlertItem> items) {
        App app = (App) application;
        if (app.isVisible()) {
            //return;
        }
        if (DataUtil.isEmpty(items)) {
            return;
        }

        String title = TextUtil.getString(app, R.string.notify_title_price_alert);
        StringBuilder message = new StringBuilder();
        Class<?> targetClass = null;
        UiTask<Coin> task = null;

        for (CoinAlertItem item : items) {
            if (message.length() > 0) {
                message.append(Constants.Sep.COMMA_SPACE);
            }
            Coin coin = item.getCoin();
            CoinAlert alert = item.getItem();
            message.append(item.getCoin().getSymbol());
            if (alert.hasPriceUp()) {
                message
                        .append(Constants.Sep.SPACE)
                        .append(Constants.Sep.UP)
                        .append(alert.getPriceUp());
            }
            if (alert.hasPriceDown()) {
                message
                        .append(Constants.Sep.SPACE)
                        .append(Constants.Sep.DOWN)
                        .append(alert.getPriceDown());
            }

            targetClass = HomeActivity.class;
            task = new UiTask<>(false, UiType.COIN, UiSubtype.VIEW, coin, null);
        }

        notify.showNotification(
                title,
                message.toString(),
                R.drawable.ic_notification,
                Constants.Notify.ALERT_ID,
                Constants.Notify.ALERT_CHANNEL_ID,
                targetClass,
                task
        );

        getApp().throwAnalytics(Constants.Event.NOTIFICATION, Constants.Companion.notifyAlertCoin(application));
    }

    private void postResultNews(List<NewsItem> items) {
        App app = (App) application;
        if (app.isVisible()) {
            //return;
        }
        if (DataUtil.isEmpty(items)) {
            return;
        }

        NewsItem news = items.get(0);
        String title = TextUtil.getString(app, R.string.notify_title_news);
        StringBuilder message = new StringBuilder();
        message.append(news.getItem().getBody());

        Class<?> targetClass = HomeActivity.class;
        UiTask<News> task = new UiTask<>(true, UiType.NEWS, UiSubtype.VIEW, news.getItem(), news.getItem().getUrl());

        notify.showNotification(
                title,
                message.toString(),
                R.drawable.ic_notification,
                Constants.Notify.NEWS_ID,
                Constants.Notify.NEWS_CHANNEL_ID,
                targetClass,
                task);
        getApp().throwAnalytics(Constants.Event.NOTIFICATION, Constants.Companion.notifyNews(application));
    }


    private void postFailed(Throwable error) {

    }

    private boolean isProfitable(Coin coin) {
        Quote quote = coin.getQuote(Currency.USD);
        return quote.getHourChange() > 0.0f || quote.getDayChange() > 0.0f || quote.getWeekChange() > 0.0f;
    }

    private boolean isAlertable(Currency currency, CoinAlert alert) {
        Coin coin = repo.getItemIf(CoinSource.CMC, currency, alert.getId());
        if (coin == null) {
            return false;
        }
        Quote quote = coin.getQuote(currency);
        if (alert.hasPriceUp() && quote.getPrice() > alert.getPriceUp()) {
            return true;
        }
        if (alert.hasPriceDown() && quote.getPrice() > alert.getPriceDown()) {
            return true;
        }
        return false;
    }
}

File: android/word/src/main/kotlin/com/dreampany/word/vm/NotifyViewModel.kt
package com.dreampany.word.vm

import android.app.Application
import com.dreampany.framework.api.notify.NotifyManager
import com.dreampany.language.Language
import com.dreampany.framework.data.model.State
import com.dreampany.framework.data.source.repository.StateRepository
import com.dreampany.framework.misc.AppExecutors
import com.dreampany.framework.misc.ResponseMapper
import com.dreampany.framework.misc.RxMapper
import com.dreampany.framework.misc.exception.EmptyException
import com.dreampany.framework.util.DataUtil
import com.dreampany.framework.util.TextUtil
import com.dreampany.framework.util.TimeUtil
import com.dreampany.network.manager.NetworkManager
import com.dreampany.translation.data.source.repository.TranslationRepository
import com.dreampany.word.R
import com.dreampany.word.app.App
import com.dreampany.word.data.enums.ItemState
import com.dreampany.word.data.enums.ItemSubtype
import com.dreampany.word.data.enums.ItemType
import com.dreampany.word.data.misc.StateMapper
import com.dreampany.word.data.misc.WordMapper
import com.dreampany.word.data.model.Word
import com.dreampany.word.data.source.pref.Pref
import com.dreampany.word.data.source.repository.WordRepository
import com.dreampany.word.misc.Constants
import com.dreampany.word.ui.activity.NavigationActivity
import com.dreampany.word.ui.enums.UiSubtype
import com.dreampany.word.ui.enums.UiType
import com.dreampany.word.ui.model.UiTask
import com.dreampany.word.ui.model.WordItem
import io.reactivex.Maybe
import timber.log.Timber
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Created by roman on 2019-07-09
 * Copyright (c) 2019 bjit. All rights reserved.
 * hawladar.roman@bjitgroup.com
 * Last modified $file.lastModified
 */

@Singleton
class NotifyViewModel
@Inject constructor(
    val application: Application,
    val rx: RxMapper,
    val ex: AppExecutors,
    val rm: ResponseMapper,
    val network: NetworkManager,
    val pref: Pref,
    val notify: NotifyManager,
    val stateMapper: StateMapper,
    val stateRepo: StateRepository,
    val wordMapper: WordMapper,
    val wordRepo: WordRepository,
    val translationRepo: TranslationRepository
) {

    fun notifyIf() {
        notifySync()
    }

    fun clearIf() {

    }

    private fun notifySync() {
        if (!TimeUtil.isExpired(pref.getLastWordSyncTime(), Constants.Delay.WordSyncTimeMS)) {
            return
        }
        pref.commitLastWordSyncTime()
        Timber.v("Fire Notification")
        val disposable = rx
            .backToMain(getSyncWordItemRx())
            .subscribe({ this.postResult(it) }, { this.postFailed(it) })
    }

    private fun getSyncWordItemRx(): Maybe<WordItem> {
        return Maybe.create { emitter ->
            //find raw word item
            //val stateCount = stateRepo.getCount(ItemType.WORD.name, ItemSubtype.DEFAULT.name, ItemState.RAW.name)
            val state = getState(ItemType.WORD, ItemSubtype.DEFAULT, ItemState.RAW)
            if (state == null || emitter.isDisposed) {
                return@create
            }
            Timber.v("Statue %s", state.toString())
            var item = wordMapper.toItemFromState(state, wordRepo)
            item = getItemIf(item)
            if (item != null) {
                pref.setRecentWord(item)
            }
            val source = Language.ENGLISH.code
            val target = pref.getLanguage(Language.ENGLISH).code
            val result = getItem(item, source, target, true);
            if (DataUtil.isEmpty(result)) {
                emitter.onError(EmptyException())
            } else {
                emitter.onSuccess(result)
            }
        }
    }

    private fun postResult(item: WordItem) {
        val app = application as App
        if (app.isVisible()) {
            //return;
        }

        val title = TextUtil.getString(app, R.string.notify_title_word_sync)
        var message: String? = if (!DataUtil.isEmpty(item.translation))
            app.getString(R.string.notify_word_translation_format, item.item.id, item.translation)
        else app.getString(R.string.notify_word_format, item.item.id, item.item.getPartOfSpeech())
        var targetClass: Class<*> = NavigationActivity::class.java

        val task = UiTask<Word>(false, UiType.WORD, UiSubtype.VIEW, item.item, null)

        notify.showNotification(title!!, message!!, R.drawable.ic_notification, targetClass, task)
        app.throwAnalytics(
            Constants.Event.NOTIFICATION,
            Constants.notifyWordSync(application)
        )
    }

    private fun postFailed(error: Throwable) {
        Timber.v(error)
    }

    fun getState(type: ItemType, subtype: ItemSubtype, state: ItemState): State? {
        val state = stateRepo.getItem(type.name, subtype.name, state.name)
        return state
    }

    fun hasState(word: Word, subtype: ItemSubtype, state: ItemState): Boolean {
        return stateRepo.getCountById(word.id, ItemType.WORD.name, subtype.name, state.name) > 0
    }

    private fun getItem(word: Word, source: String, target: String, fully: Boolean): WordItem {
        val item = WordItem.getItem(word)
        item.item = word
        adjustTranslate(item, source, target)
        return item
    }

    fun putState(word: Word, subtype: ItemSubtype, state: ItemState): Long {
        val s = State(word.id, ItemType.WORD.name, subtype.name, state.name)
        s.time = TimeUtil.currentTime()
        return stateRepo.putItem(s)
    }

    fun putItem(word: Word, subtype: ItemSubtype, state: ItemState): Long {
        var result = wordRepo.putItem(word)
        if (result != -1L) {
            result = putState(word, subtype, state)
        }
        return result
    }

    fun getItemIf(word: Word): Word? {
        var result = getRoomItemIf(word)
        if (result == null) {
            result = getFirestoreItemIf(word)
        }
        if (result == null) {
            result = getRemoteItemIf(word)
        }
        return result
    }

    private fun getRoomItemIf(word: Word): Word? {
        return if (!hasState(word, ItemSubtype.DEFAULT, ItemState.FULL)) {
            null
        } else wordRepo.getRoomItem(word.id, true)
    }

    private fun getFirestoreItemIf(word: Word): Word? {
        val result = wordRepo.getFirestoreItem(word.id, true)
        if (result != null) {
            Timber.v("Firestore result success")
            this.putItem(result, ItemSubtype.DEFAULT, ItemState.FULL)
        }
        return result
    }

    private fun getRemoteItemIf(word: Word): Word? {
        val result = wordRepo.getRemoteItem(word.id, true)
        if (result != null) {
            this.putItem(result, ItemSubtype.DEFAULT, ItemState.FULL)
            wordRepo.putFirestoreItem(result)
        }
        return result
    }

    private fun adjustTranslate(item: WordItem, source: String, target: String) {
        if (source.equals(target)) {
            return
        }
        var translation: String? = null
        if (item.hasTranslation(target)) {
            translation = item.getTranslationBy(target)
        } else {
            val textTranslation =
                translationRepo.getItem(item.item.id, source, target)
            Timber.v("Translation %s - %s", item.item.id, translation)
            textTranslation?.let {
                item.addTranslation(target, it.output)
                translation = it.output
            }
        }
        item.translation = translation
    }
}